Allow custom health configuration to be deployed as annotations

aljesusg opened the pull request:
Fix https://github.com/kiali/kiali/issues/3332
Fix https://github.com/kiali/kiali/issues/3435

# Support Health Config Annotations

- [X] Service
- [X] Workload

# Graph side
- [X] Service
- [x] Workload

aljesusg commented:
Some concerns.

# Workload case

I need to check all the pods start with workload name requested. After that check all annotations (could be different, example : manual update and have a reviews-v1-323232 and reviews-v1-4343 with different annotation) What to do?

# App case

I think that we should remove this improvement for this kind resource

# Cache

I saw the cache work for namespaces. I was thinking about to store for a specific resource the configuration to avoid to fetch it each time. Is the cache refreshing when annotations change in kubernetes?

cc @lucasponce @jotak

jotak commented:
> Some concerns.
> Workload case
> 
> I need to check all the pods start with workload name requested.

why checking pod name and workload name? why not workload-versus-workload? (see other point below maybe it's related)

> After that check all annotations (could be different, example : manual update and have a reviews-v1-323232 and reviews-v1-4343 with different annotation) What to do?

I think we can assume pods within the same workload share the same annotations. Actually I think we don't have other choice than to assume that (we won't have health per pod anyway). So, I would say just picking just the first pod is fine :)

> App case
> 
> I think that we should remove this improvement for this kind resource

I see why you say that... the annotation could only be set a the workload level, but the health that we fetch for apps are all merged into a single RequestHealth object. Perhaps we should change the model here, to allow per-workload config within app health, but it's fine to postpone that for later, one step after another.

> Cache
> 
> I saw the cache work for namespaces. I was thinking about to store for a specific resource the configuration to avoid to fetch it each time. Is the cache refreshing when annotations change in kubernetes?
> 
> cc @lucasponce @jotak

aljesusg commented:
> > Some concerns.
> > Workload case
> > I need to check all the pods start with workload name requested.
> 
> why checking pod name and workload name? why not workload-versus-workload? (see other point below maybe it's related)
> 
> > After that check all annotations (could be different, example : manual update and have a reviews-v1-323232 and reviews-v1-4343 with different annotation) What to do?
> 
> I think we can assume pods within the same workload share the same annotations. Actually I think we don't have other choice than to assume that (we won't have health per pod anyway). So, I would say just picking just the first pod is fine :)

Yes...there is not another way that assume that...I hope that kubernetes will update it for all the pods

> 
> > App case
> > I think that we should remove this improvement for this kind resource
> 
> I see why you say that... the annotation could only be set a the workload level, but the health that we fetch for apps are all merged into a single RequestHealth object. Perhaps we should change the model here, to allow per-workload config within app health, but it's fine to postpone that for later, one step after another.

Yes..I'll start adding this feature to service and workload and continue in another step with app.

> 
> > Cache
> > I saw the cache work for namespaces. I was thinking about to store for a specific resource the configuration to avoid to fetch it each time. Is the cache refreshing when annotations change in kubernetes?
> > cc @lucasponce @jotak

lucasponce commented:
> I need to check all the pods start with workload name requested. After that check all annotations (could be different, example : manual update and have a reviews-v1-323232 and reviews-v1-4343 with different annotation) What to do?

I'd check only the workload / controller template (or pod in the corner case that there is no controller).

> I think that we should remove this improvement for this kind resource

It's fine, also, Application doesn't exist as a strong entity, so, it's fine to not consider it in a first step. If we cover the Workload/Services sounds a good progress. We can see on testing if there is some clash between workload and app.

lucasponce commented:
@aljesusg I guess it's still in active development and perhaps it's earlier to jump into the code or doing more testing.
Just ping when you think it's ready to further review or local testing and we can help there.

aljesusg commented:
Ok I added a new appender to return the annotations for workloads graph type in first step, I am going to check the changes in the UI and we can think about to add services/apps

aljesusg commented:
Ok @lucasponce now we can see anntoations in graph endpoint. I am filtering the annotations to return only rates

Annotation
```json
{"kiali": {"health_config": {"rate":[{"code": "^4\\d\\d$", "protocol": "http",
      "degraded": 5, "failure": 10}]}}
```

Example node
```json
{
        "data": {
          "id": "e8a4c5a8a5a937ec63d1da940d4b68a1",
          "nodeType": "service",
          "namespace": "bookinfo",
          "app": "reviews",
          "service": "reviews",
          "annotation": [
            {
              "code": "^4\\d\\d$",
              "degraded": 5,
              "failure": 10,
              "protocol": "http"
            }
          ],
          "destServices": [
            {
              "namespace": "bookinfo",
              "name": "reviews"
            }
          ]
        }
      }
```

I don't know if we should call **kialiAnnotation** instead **annotation** in the return object

aljesusg commented:
# How to test this 

We can check for example in service reviews that the health is green.

We are going to force this having a tolerance for all codes

```bash
kubectl annotate --overwrite svc reviews kiali="{\"health_config\":{\"rate\": [{\"code\": \"^.*$\", \"failure\": 20, \"degraded\": 11, \"protocol\": \"http\"}]}}" -n bookinfo
```
## Screenshot
![Screenshot from 2020-11-13 09-36-37](https://user-images.githubusercontent.com/3019213/99046946-c6c2af80-2593-11eb-8b23-c7442f568677.png)

aljesusg commented:
# Fast example

Deploy error-rates with this config in operator

```bash
{
  "healthConfig": {
    "rate": [
      {
        "namespace": "/alpha/",
        "kind": "/app/",
        "name": "/x-server/",
        "tolerance": [
          {
            "code": "/[45]\\d[^\\D4]/",
            "degraded": 0,
            "failure": 10,
            "protocol": "/http/",
            "direction": "/.*/"
          }
        ]
      }    
    ]
  }
}
```

We'll see the next picture in the graph
![Screenshot from 2020-11-16 10-19-05](https://user-images.githubusercontent.com/3019213/99234718-2f5c9700-27f5-11eb-8368-36c01f0e2eee.png)
![Screenshot from 2020-11-16 10-33-19](https://user-images.githubusercontent.com/3019213/99236195-2d93d300-27f7-11eb-8ebe-9f9c8d39aae3.png)


Where X-server is in green status but we can check that we have 9.0% of 400 error code

![Screenshot from 2020-11-16 10-20-02](https://user-images.githubusercontent.com/3019213/99234807-4e5b2900-27f5-11eb-9c9c-8ba49cbdc57f.png)

We want to show this error for x-server when the error rate is >5%, so we need to add the next annotation 

kubectl annotate --overwrite svc x-server kiali="{\"health_config\":{\"rate\": [{\"code\": \"^4\\\d\\\d$\", \"failure\": 20, \"degraded\": 5, \"protocol\": \"http\"}]}}" -n alpha


and now we can check the graph with this degraded status

![Screenshot from 2020-11-16 10-32-47](https://user-images.githubusercontent.com/3019213/99236212-3389b400-27f7-11eb-8ac6-e5b85a0a1c57.png)
![Screenshot from 2020-11-16 10-32-57](https://user-images.githubusercontent.com/3019213/99236223-35ec0e00-27f7-11eb-97fc-2a4baab28deb.png)

hhovsepy commented:
Have increased the degraded rates for service and workload to 10%, but still it shows warning icon for both:
kubectl annotate --overwrite svc y-server kiali="{"health_config":{"rate": [{"code": "^4\\\d\\\d$", "failure": 20, "degraded": 10, "protocol": "http"}]}}" -n alpha
kubectl annotate --overwrite deployment b-client kiali="{"health_config":{"rate": [{"code": "^4\\\d\\\d$", "failure": 20, "degraded": 10, "protocol": "http"}]}}" -n beta

![Screenshot from 2020-11-20 09-08-39](https://user-images.githubusercontent.com/604313/99776340-f41dd900-2b10-11eb-914e-862d181cba37.png)
![Screenshot from 2020-11-20 09-09-31](https://user-images.githubusercontent.com/604313/99776341-f54f0600-2b10-11eb-864b-905724e42f8e.png)

aljesusg commented:
> Have increased the degraded rates for service and workload to 10%, but still it shows warning icon for both:
> kubectl annotate --overwrite svc y-server kiali="{"health_config":{"rate": [{"code": "^4\\d\\d$", "failure": 20, "degraded": 10, "protocol": "http"}]}}" -n alpha
> kubectl annotate --overwrite deployment b-client kiali="{"health_config":{"rate": [{"code": "^4\\d\\d$", "failure": 20, "degraded": 10, "protocol": "http"}]}}" -n beta
> 
> ![Screenshot from 2020-11-20 09-08-39](https://user-images.githubusercontent.com/604313/99776340-f41dd900-2b10-11eb-914e-862d181cba37.png)
> ![Screenshot from 2020-11-20 09-09-31](https://user-images.githubusercontent.com/604313/99776341-f54f0600-2b10-11eb-864b-905724e42f8e.png)

The problem is that in that PR we are not throwing errors, we are working in another PR.

So You can check if the config is correct if the health endpoint return that config. In that case /api/namespaces/alpha/services/y-server/health not return the configuration. If you go to /api/namespaces/alpha/services/y-server you can check the kiali annotation is 
```json
{health_config:{rate: [{code: ^4\\d\\d, failure: 20, degraded: 10, protocol: http}]}}
```
and should be
```json
{\"health_config\":{\"rate\": [{\"code\": \"^4\\d\\d\", \"failure\": 20, \"degraded\": 10, \"protocol\": \"http\"}]}}
```


The cli command should be
```bash
kubectl annotate --overwrite svc x-server kiali="{\"health_config\": {\"rate\": [{\"code\": \"^4\\\d\\\d$\", \"degraded\": 10, \"failure\": 20, \"protocol\": \"http\"}]}}" -n alpha
```

hhovsepy reviewed (dismissed):
Thanks @aljesusg verified now.

lucasponce commented:
Hi @aljesusg would you mind to rebase this and the UI one ?

I see the branch is a little bit old
![image](https://user-images.githubusercontent.com/1662329/99959850-08213f00-2d8c-11eb-9e54-7ed429436ea4.png)

I'm going to start to test & review it.

lucasponce commented:
We have discussed a couple of suggestions on this feature:
- Adjust the annotation format in a simplified structure.
- Introduce some "sugar" syntax for the code.

Inspired in [1] this annotation could be something named like:

```
health.kiali.io/rate
```

In that way, this syntax allows to extend and add more categories in the future and it's sync-ed with istio ones.

Also, the tolerance rate is defined with the following structure:

```
	Code      string  `yaml:"code,omitempty" json:"code"`
	Degraded  float32 `yaml:"degraded,omitempty" json:"degraded"`
	Failure   float32 `yaml:"failure,omitempty" json:"failure"`
	Protocol  string  `yaml:"protocol,omitempty" json:"protocol"`
	Direction string  `yaml:"direction,omitempty" json:"direction"`
```

In a configuration for the namespace is used like:

```
    - namespace: "alpha"
      tolerance:
        - code: "404"
          failure: 10
          protocol: "http"
        - code: "[45]\\d[^4]"
          protocol: "http"
```

But this tolerance could be represented in an annotation for a service / workload like:

```
health.kiali.io/rate=404,0,10;[45]x[^4],10,20
```

Which is better readable for a user.

Two details:
- As it happens with the global config, when a field is not present it takes the default value.
- For **code** regex we expect to have only numeric values, we could introduce a "sugar syntax" with x or X to represent `\d`, we'll reduce the number of parsing problems and also it's readable for the user expression like 4xx or 5xx (or 4XX, 5XX) in this context.

This "sugar syntax" could be applied to the general config as well:

```
    - namespace: "alpha"
      tolerance:
        - code: "404"
          failure: 10
          protocol: "http"
        - code: "[45]X[^4]"
          protocol: "http"
```

I think this is going to trigger some refactoring in the current PR but it shouldn't change the result of the feature.


[1] https://istio.io/latest/docs/reference/config/annotations/

aljesusg commented:
I did the changes in backend and UI side @lucasponce. I'm going to continue working in the UI.

Another thing....For workload we are fetching the annotations of the first pod....maybe we should fetch the deployment instead this? What do you think @jotak @lucasponce ?

lucasponce commented:
> Another thing....For workload we are fetching the annotations of the first pod....maybe we should fetch the deployment instead this? What do you think @jotak @lucasponce ?

Yes, I have that comment in my list for the review, but as we were discussing more the changes in the feature I didn't want to mix all feature + low level code reviews.

I'll resume the review and I'll add more comments.

aljesusg commented:
> > Another thing....For workload we are fetching the annotations of the first pod....maybe we should fetch the deployment instead this? What do you think @jotak @lucasponce ?
> 
> Yes, I have that comment in my list for the review, but as we were discussing more the changes in the feature I didn't want to mix all feature + low level code reviews.
> 
> I'll resume the review and I'll add more comments.

I thought that deployment change propagate this changes....:(

lucasponce commented on models/workload.go:
```diff
@@ -127,6 +131,10 @@ func (workload *WorkloadListItem) ParseWorkload(w *Workload) {
 	workload.PodCount = len(w.Pods)
 	workload.AdditionalDetailSample = w.AdditionalDetailSample
 
+	if len(w.Pods) > 0 {
```
I'd expect the annotation on the workload, no need to get it from the Pod.
Just place the annotation inside the `parseObjectMeta` and that's good.
Note, that all workload annotations manipulation are located there.

jotak replied:
I don't know, the advantage of checking at pod level is that, in the end, there's always a pod ; whereas deployments are not mandatory, pods can result from various kind of controllers.
Like istio annotations in general are taken from pods, so I thought we could follow the same pattern.

jotak replied:
(see how many istio annotations are read from pods: https://istio.io/latest/docs/reference/config/annotations/ )

aljesusg replied:
In fact I checked that yesterday thanks @jotak.

In my last commit I am filtering by kiali Annotations.

lucasponce replied:
Deployments scaled to 0 are possible, a corner case but possible.
For this particular code, the suggestion is to move it into the parseObjectMeta() method that is performing a similar operation.
In that case, it will cover all cases, Pod included.

lucasponce commented on models/health.go:
```diff
@@ -1,9 +1,14 @@
 	"github.com/prometheus/common/model"
 )
 
+const HealthConfiguration = "health_config"
```
Unused

lucasponce commented on models/health.go:
```diff
@@ -1,9 +1,14 @@
 )
 
+const HealthConfiguration = "health_config"
+const HealthConfigurationRates = "rate"
```
Unused

lucasponce commented on models/health.go:
```diff
@@ -157,3 +163,10 @@ func (ps ProxyStatus) IsSynced() bool {
 }
+
+func GetHealthConfiguration(annotations map[string]string) (string, error) {
+	if kialiConf, ok := annotations["health.kiali.io/rate"]; ok {
```
Naming is hard and opnionated, just thinking in future maintenance, to me it would be more clear "kialiConfig" -> "healthConfig".
It's a minor comment.

lucasponce commented on models/health.go:
```diff
@@ -89,6 +94,7 @@ type ProxyStatus struct {
 type RequestHealth struct {
 	Inbound  map[string]map[string]float64 `json:"inbound"`
 	Outbound map[string]map[string]float64 `json:"outbound"`
+	Config   string                        `json:"config"`
```
This should be `omitempty` as this is optional and only used on the case where user annotates a specific Service and/or Workload.

lucasponce replied:
I'm reviewing the appender and looking that code I guess we should call this parameter "healthConfig", also here and in the graph to get consistency.
Rationale is that /health endpoint and /graph fetch the info needed to calculate health: rate request, if "healthConfig" is present in both, then instead to use the global "healthConfig" config it personalizes for this item (service/workload) using this info.

lucasponce commented on models/health.go:
```diff
@@ -157,3 +163,10 @@ func (ps ProxyStatus) IsSynced() bool {
 	return componentStatus == "Synced"
 }
+
+func GetHealthConfiguration(annotations map[string]string) (string, error) {
```
We reserve the "Get*" verbs for backend calls, for these kind of operations a "Parse*" / "Cast*" fits better because it's just extracting data from one format to another.

aljesusg replied:
I nned to remove this method is not used :(

lucasponce commented on models/health.go:
```diff
@@ -157,3 +163,10 @@ func (ps ProxyStatus) IsSynced() bool {
+	if kialiConf, ok := annotations["health.kiali.io/rate"]; ok {
+		return kialiConf, nil
+	}
+	return "", fmt.Errorf("No Kiali Health configuration in Annotations")
```
I don't think you should return an error here.
Not every Service would expect annotations.
So, I would send just empty string, then returning error may make no sense on this method (and all propagations).

lucasponce commented on business/health.go:
```diff
@@ -291,9 +293,23 @@ func fillWorkloadRequestRates(allHealth models.NamespaceWorkloadHealth, rates mo
 		rqHealth.AggregateInbound(sample)
 	}
+
+	svc, err := in.k8s.GetService(namespace, service)
```
Ok, there is a missing case in the kialiCached that I missed.
This query is ok, but it should be done via cache.
No changes for the moment as this requires more changes.

I'll send a followup PR https://github.com/kiali/kiali/issues/3467 and I will update this later.

lucasponce replied:
Note, that PR #3467 is merged so business package can get this from cached.
i.e. used from validations
```
svc, err := in.businessLayer.Svc.getService(namespace, service)
```

lucasponce commented on business/health.go:
```diff
@@ -312,11 +328,26 @@ func (in *HealthService) getAppRequestsHealth(namespace, app, rateInterval strin
+	if err != nil {
+		return rqHealth, err
+	}
+	if len(w.Pods) > 0 {
```
Annotation should be provided by the model.Workload instead to get parsed from the Pod.

lucasponce commented on business/health.go:
```diff
@@ -312,11 +328,26 @@ func (in *HealthService) getAppRequestsHealth(namespace, app, rateInterval strin
+		var errh error
+		rqHealth.Config, errh = models.GetHealthConfiguration(w.Pods[0].Annotations)
+		if errh != nil {
+			log.Debugf("Error Getting Kiali Health Configuration. %s", errh.Error())
```
This sounds like a debug, but it's not an error, because the normal scenario is that annotations are not present.

lucasponce commented on graph/config/cytoscape/cytoscape.go:
```diff
@@ -59,26 +59,27 @@ type NodeData struct {
+	HasCB            bool                `json:"hasCB,omitempty"`            // true (has circuit breaker) | false
+	HasMissingSC     bool                `json:"hasMissingSC,omitempty"`     // true (has missing sidecar) | false
+	HasVS            bool                `json:"hasVS,omitempty"`            // true (has route rule) | false
+	HealthAnnotation []interface{}       `json:"healthAnnotation,omitempty"` // requested healthAnnotation for NodeType
```
I guess "Annotation" shouldn't belong to this domain, I guess it's better to name it "HealthConfig" and I think type should be a `string` if I'm not wrong and I missed something.

lucasponce commented on graph/meta.go:
```diff
@@ -13,6 +13,7 @@ func NewMetadata() Metadata {
 
 // Metadata keys to be used instead of literal strings
 const (
+	Annotation      MetadataKey = "annotation"
```
I have feeling using "annotation" naming will be confusing, "healthConfig" would help with the semantic.

lucasponce commented on graph/telemetry/istio/appender/appender.go:
```diff
@@ -22,6 +22,8 @@ func ParseAppenders(o graph.TelemetryOptions) []graph.Appender {
 	if !o.Appenders.All {
 		for _, appenderName := range o.Appenders.AppenderNames {
 			switch appenderName {
+			case AnnotationAppenderName:
```
The previous comment about naming would propagate into these changes as well.

lucasponce commented on graph/telemetry/istio/appender/health_annotation.go:
```diff
@@ -0,0 +1,75 @@
+	"github.com/kiali/kiali/models"
+)
+
+const AnnotationAppenderName = "healthAnnotation"
```
"healthAnnotation" is a little confusing to me, I'd prefer a "healthConfig" naming.
Another consideration is that this appender is pretty close to the sidecar one, and perhaps both could be merged together.

But I'd let @jshaughn to comment here.

aljesusg replied:
I am goign to replace this by the new syntax "health.kiali.io/rate" , if we add more annotations we can change this to "kiali annotations" to return all annotations like kiali.io

lucasponce replied:
The suggestion is to use a "healthConfiguration" appender instead of "healthAnnotations".
When we cover a more generic case we can reconsider, but "healthAnnotation" is a little bit confusing.

lucasponce reviewed (changes requested):
Conceptually feature is clear:
- Both sources of info used to calculate health (/graph and /health) can be augmented with a specific "healthConfig" info that will be used to compute health instead of using the global config defined. This optional "healthConfig" can be added as part of a kiali annotation in Workloads and/or Services.

Some requests about code organization and naming.

Also, this PR should wait to https://github.com/kiali/kiali/issues/3467 as it's missing to use the cache for Service Detail in order to not invoke unnecessary calls to the backend.

jshaughn commented on graph/telemetry/istio/appender/kiali_annotation.go:
```diff
@@ -0,0 +1,69 @@
+	"github.com/kiali/kiali/models/annotation"
+)
+
+const AnnotationAppenderName = "kialiAnnotation"
```
This is similar to the Sidecars appender and could feasibly be combined into that appender, but there are a few differences and I'm not opposed to adding a new appender specifically to handle these config overrides.   As mentioned above, I do think the name should change to better reflect what it is doing.  Assuming the proposed name change:

`const HealthConfigAppenderName = "healthConfig"`

jshaughn commented on graph/telemetry/istio/appender/kiali_annotation.go:
```diff
@@ -0,0 +1,69 @@
+const AnnotationAppenderName = "kialiAnnotation"
+
+// Name: annotation
+type AnnotationAppender struct{}
```
This would change to `HealthConfigAppender`.  Also, the appenders typically have a fairly robust comment above type decl.

Also, before a merge we'd need test code for this appender.

jshaughn commented on graph/telemetry/istio/appender/kiali_annotation.go:
```diff
@@ -0,0 +1,69 @@
+		graph.CheckError(err)
+		namespaceInfo.Vendor[workloadListKey] = &workloadList
+	}
+
```
You don't want to get the ServiceDefinitionList every time, instead you want to grab the cached version that presumably would have been established by `IstioAppender`.  So insert here:

```
	if getServiceDefinitionList(namespaceInfo) == nil {
		sdl, err := globalInfo.Business.Svc.GetServiceDefinitionList(namespaceInfo.Namespace)
		graph.CheckError(err)
		namespaceInfo.Vendor[serviceDefinitionListKey] = sdl
	}
```

jshaughn commented on graph/telemetry/istio/appender/kiali_annotation.go:
```diff
@@ -0,0 +1,69 @@
+	a.applyAnnotationPresence(trafficMap, namespaceInfo)
+}
+
+func (a *AnnotationAppender) applyAnnotationPresence(trafficMap graph.TrafficMap, namespaceInfo *graph.AppenderNamespaceInfo) {
```
just a name change here...

jshaughn commented on graph/telemetry/istio/appender/kiali_annotation.go:
```diff
@@ -0,0 +1,69 @@
+
+func (a *AnnotationAppender) applyAnnotationPresence(trafficMap graph.TrafficMap, namespaceInfo *graph.AppenderNamespaceInfo) {
+	for _, n := range trafficMap {
+		// Skip the check if this node is outside the requested namespace, we limit badging to the requested namespaces
+		if n.Namespace != namespaceInfo.Namespace {
+			continue
+		}
```
I think this check is valid as we onbly have the workloads for the current namespace.  Juts need to fix the comment.

jshaughn commented on graph/telemetry/istio/appender/kiali_annotation.go:
```diff
@@ -0,0 +1,69 @@
+			continue
+		}
+
+		// We whitelist istio components because they may not report telemetry using injected sidecars.
+		if config.IsIstioNamespace(n.Namespace) {
+			continue
+		}
```
This I'm really not sure about.  We probably want to delete these lines as it seems feasible that a user could annotate workloads in the control plane namespace.

jshaughn commented on graph/telemetry/istio/appender/kiali_annotation.go:
```diff
@@ -0,0 +1,69 @@
+		}
+
+		// dead nodes tell no tales (er, have no pods)
+		if isDead, ok := n.Metadata[graph.IsDead]; ok && isDead.(bool) {
+			continue
+		}
```
Dead nodes have no pods but I think it's possible they have workloads.  I think we should delete these lines.

aljesusg replied:
In this case we can remove them, because for workload we are fetching the annotations of their first pod, so if they have not pods so not makes sense.

jshaughn commented on graph/telemetry/istio/appender/kiali_annotation.go:
```diff
@@ -0,0 +1,69 @@
+			continue
+		}
+
+		// get the workloads for the node and check to see if they have sidecars. Note that
+		// if there are no workloads/pods we don't flag it as missing sidecars.  No pods means
+		// no missing sidecars.  (In most cases this means it was flagged as dead, and handled above)
```
Fix the comment

jshaughn commented on graph/telemetry/istio/appender/kiali_annotation.go:
```diff
@@ -0,0 +1,69 @@
+		switch n.NodeType {
+		case graph.NodeTypeWorkload:
+			if workload, found := getWorkload(n.Workload, namespaceInfo); found {
+				annotations := annotation.GetAnnotation(workload.Annotations, annotation.GetKialiAnnotation())
```
I think there will be name changes here based on other comments.  But if not we want to only send back healthConfig overrides.  Again, I'd like to see Kiali-level typing as opposed to just the Go map.

jshaughn reviewed (changes requested):
@aljesusg After a short meeting with @lucasponce I now understand better the approach of this feature, and I do see the need for the additional appender support.  Because we still generate the health determination on the client, and this is not the static config we pull already from the server, we do need to look for the config overrides at graph-generation-time.  The good news is that the additional overhead should be negligible, as we already pull the information needed in other, non-optional, appenders.

It looks like you found the necessary touch points in the code, and it looks pretty close, but I did have a few comments.  Also, when the appender name is decided you will want to add the appender name to line 172 of doc.go.

aljesusg commented on graph/config/cytoscape/cytoscape.go:
```diff
@@ -71,6 +71,7 @@ type NodeData struct {
 	HasCB           bool                `json:"hasCB,omitempty"`           // true (has circuit breaker) | false
 	HasMissingSC    bool                `json:"hasMissingSC,omitempty"`    // true (has missing sidecar) | false
 	HasVS           bool                `json:"hasVS,omitempty"`           // true (has route rule) | false
+	KialiAnnotation map[string]string   `json:"kialiAnnotation,omitempty"` // requested kialiAnnotation for NodeType
```
understood, I was thinking more in let this for future enhancements but you're right...our first step is only focus in health. I'll address the comments, thanks @jshaughn

aljesusg replied:
instead to HealthConfig this should be a []Tolerance, it makes no sense set the full configuration for a specific node.

jshaughn replied:
I have a inline response to this in my next review.

jshaughn commented on graph/config/cytoscape/cytoscape.go:
```diff
@@ -71,6 +71,7 @@ type NodeData struct {
 	HasCB           bool                `json:"hasCB,omitempty"`           // true (has circuit breaker) | false
 	HasMissingSC    bool                `json:"hasMissingSC,omitempty"`    // true (has missing sidecar) | false
 	HasVS           bool                `json:"hasVS,omitempty"`           // true (has route rule) | false
+	HasHealthConfig map[string]string   `json:"HasHealthConfig,omitempty"` // set to the health config override
```
three things:
- json should be `hasHealthConfig`
- maintain alphebetical (move up after HasCB)
- recommended: I still think map[string]string deserves to be typed.  I had suggested `HealthConfig`, you suggest ` []Tolerance`.  But for me `[]Tolerance` introduces another term that then needs definition.    Remember that the type declared here is just local to this Cytoscape config generator, declared above with the other types.  So I think something like this would be fine (update the comment as you see fit):
```
// HealthConfig is a flexible map to pass available health information, tolerances, etc, to be interpreted by the client.
type HealthConfig map[string][string]
```

aljesusg replied:
understood. Thanks

jshaughn commented on graph/meta.go:
```diff
@@ -13,6 +13,7 @@ func NewMetadata() Metadata {
 
 // Metadata keys to be used instead of literal strings
 const (
+	HasHealthConfig MetadataKey = "hasHealthConfig"
```
Name change is good, please fix alphabetical ordering.

jshaughn commented on graph/telemetry/istio/appender/appender.go:
```diff
@@ -22,6 +22,8 @@ func ParseAppenders(o graph.TelemetryOptions) []graph.Appender {
 	if !o.Appenders.All {
 		for _, appenderName := range o.Appenders.AppenderNames {
 			switch appenderName {
+			case HealthConfigAppenderName:
+				requestedAppenders[HealthConfigAppenderName] = true
```
Name change is good. Again, maintain alphabetical ordering where it is being used, it helps in code readability.

jshaughn commented on graph/telemetry/istio/appender/health_config.go:
```diff
@@ -0,0 +1,52 @@
+		graph.CheckError(err)
+		namespaceInfo.Vendor[workloadListKey] = &sdl
+	}
+
```
Still missing the getServiceDefinitionList.  See my previous review comment.

aljesusg replied:
I added it in L25 @jshaughn 

https://github.com/kiali/kiali/blob/a542e95b525a6e0f903d6f236e333c818d88cc87/graph/telemetry/istio/appender/health_config.go#L25-L27

jshaughn reviewed (changes requested):
Looking better, still some changes to make, ans still need test code for the new appender.

lucasponce commented:
I'd reviewed on Monday, but is there a UI more or less in good shape to test the end-to-end cases ?
I can review the code but I'd like to start from the feature first.

lucasponce reviewed (changes requested):
It needs a counterpart in the UI updated to be reviewed together in the same context.
Otherwise, if it's merged without UI review it may require changes in backend.

PR https://github.com/kiali/kiali-ui/pull/1986 is non update since some time.

No rush, but just let reviewers when it's ready to review them together.

lucasponce commented:
@aljesusg would you mind to review the CI errors on this PR ?
Thanks

lucasponce commented:
The regular expressions used are something that I'm not 100% conviced.

As we have introduced some "sugar syntax" for these scenario, and custom Regex here will be rare, I suggest to change the defaults in something that user can read better:

Original:
```
				Tolerance: []Tolerance{
					{
						Code:      "^5\\d\\d$",
						Protocol:  "http",
						Direction: ".*",
						Failure:   10,
					},
					{
						Code:      "^4\\d\\d$",
						Protocol:  "http",
						Direction: ".*",
						Degraded:  10,
						Failure:   20,
					},
					{
						Code:      "^[1-9]$|^1[0-6]$",
						Protocol:  "grpc",
						Direction: ".*",
						Failure:   10,
					},
					{
						Code:      "^-$", // no response is indicated with a "-" code
						Protocol:  "http|grpc",
						Direction: ".*",
						Failure:   10,
					},
				},
```

Suggested:
```
				Tolerance: []Tolerance{
					{
						Code:      "5XX",
						Protocol:  "http",
						// Direction: ".*", <- An empty value could be considered as ".*"
						Failure:   10,
					},
					{
						Code:      "4XX",
						Protocol:  "http",
						// Direction: ".*", <- An empty value could be considered as ".*"
						Degraded:  10,
						Failure:   20,
					},
					{
						Code:      "^[1-9]$|^1[0-6]$",
						Protocol:  "grpc",
						// Direction: ".*", <- An empty value could be considered as ".*"
						Failure:   10,
					},
					{
						Code:      "^-$", // no response is indicated with a "-" code
						Protocol:  "http|grpc",
						// Direction: ".*", <- An empty value could be considered as ".*"
						Failure:   10,
					},
				},
```

This will also help when iterate in how to show this information to the user.

aljesusg commented:
I only applied the 'X' Sintax to the annotations not to the default configuration. I need to change more code for this let me address it.

lucasponce commented:
Also, just put on the user's role:

```
{
  "healthConfig": {
    "rate": [
      {
        "namespace": "/.*/",
        "kind": "/.*/",
        "name": "/.*/",
        "tolerance": [
          {
            "code": "/^5\\d\\d$/",
            "degraded": 0,
            "failure": 10,
            "protocol": "/http/",
            "direction": "/.*/"
          },
          {
            "code": "/^4\\d\\d$/",
            "degraded": 10,
            "failure": 20,
            "protocol": "/http/",
            "direction": "/.*/"
          },
          {
            "code": "/^[1-9]$|^1[0-6]$/",
            "degraded": 0,
            "failure": 10,
            "protocol": "/grpc/",
            "direction": "/.*/"
          },
          {
            "code": "/^-$/",
            "degraded": 0,
            "failure": 10,
            "protocol": "/http|grpc/",
            "direction": "/.*/"
          }
        ]
      }
    ]
  }
}
```

could be simplified to
```
{
  "healthConfig": {
    "rate": [
      {
        "tolerance": [
          {
            "code": "5XX",
            "degraded": 0,
            "failure": 10,
            "protocol": "http",
          },
          {
            "code": "4XX",
            "degraded": 10,
            "failure": 20,
            "protocol": "http",
          },
          {
            "code": "/^[1-9]$|^1[0-6]$/",
            "degraded": 0,
            "failure": 10,
            "protocol": "grpc",
          },
          {
            "code": "/^-$/",
            "degraded": 0,
            "failure": 10,
            "protocol": "http|grpc",
          }
        ]
      }
    ]
  }
}
```

So, when some field is not present, it could be considered empty and and empty would mean ".*" in practice, then it would simplify the read of this.

The idea here is to allow regexp, but those should be more corner cases rather than the usual scenario.

aljesusg commented:
> Also, just put on the user's role:
> 
> ```
> {
>   "healthConfig": {
>     "rate": [
>       {
>         "namespace": "/.*/",
>         "kind": "/.*/",
>         "name": "/.*/",
>         "tolerance": [
>           {
>             "code": "/^5\\d\\d$/",
>             "degraded": 0,
>             "failure": 10,
>             "protocol": "/http/",
>             "direction": "/.*/"
>           },
>           {
>             "code": "/^4\\d\\d$/",
>             "degraded": 10,
>             "failure": 20,
>             "protocol": "/http/",
>             "direction": "/.*/"
>           },
>           {
>             "code": "/^[1-9]$|^1[0-6]$/",
>             "degraded": 0,
>             "failure": 10,
>             "protocol": "/grpc/",
>             "direction": "/.*/"
>           },
>           {
>             "code": "/^-$/",
>             "degraded": 0,
>             "failure": 10,
>             "protocol": "/http|grpc/",
>             "direction": "/.*/"
>           }
>         ]
>       }
>     ]
>   }
> }
> ```
> 
> could be simplified to
> 
> ```
> {
>   "healthConfig": {
>     "rate": [
>       {
>         "tolerance": [
>           {
>             "code": "5XX",
>             "degraded": 0,
>             "failure": 10,
>             "protocol": "http",
>           },
>           {
>             "code": "4XX",
>             "degraded": 10,
>             "failure": 20,
>             "protocol": "http",
>           },
>           {
>             "code": "/^[1-9]$|^1[0-6]$/",
>             "degraded": 0,
>             "failure": 10,
>             "protocol": "grpc",
>           },
>           {
>             "code": "/^-$/",
>             "degraded": 0,
>             "failure": 10,
>             "protocol": "http|grpc",
>           }
>         ]
>       }
>     ]
>   }
> }
> ```
> 
> So, when some field is not present, it could be considered empty and and empty would mean ".*" in practice, then it would simplify the read of this.
> 
> The idea here is to allow regexp, but those should be more corner cases rather than the usual scenario.
> 
> I'm not sure if I'm missing some case, but that's the idea, to simplify this config, and trying to unify them with also the annotation one.

In fact if a field is empty Kiali will use "allMatch" with the value .*

lucasponce commented:
> In fact if a field is empty Kiali will use "allMatch" with the value .*

I haven't review the implementation details, I'm just first review this from user's point of view, then perhaps we need to update the defaults in Kiali and operator/helm etc.

aljesusg commented:
> > In fact if a field is empty Kiali will use "allMatch" with the value .*
> 
> I haven't review the implementation details, I'm just first review this from user's point of view, then perhaps we need to update the defaults in Kiali and operator/helm etc.

Yes, I remember that we added that to the doc to show the user the options in the example, it makes no sense when we have the "table" doc with all opts. For each expresion we use this method in the frontend.

So in the case that we have not a value we return allMatch

```ts
const allMatch = new RegExp('.*');
export const getExpr = (value: RegexConfig | undefined, code: boolean = false): RegExp => {
  if (value) {
    if (typeof value === 'string' && value !== '') {
      const v = value.replace('\\\\', '\\');
      return new RegExp(code? replaceXCode(v):v);
    }
    if (typeof value === 'object' && value.toString() !== '/(?:)/') {
      return value;
    }
  }
  return allMatch;
};
```

I added the condition to let user set X or x in the code option @lucasponce 
https://github.com/kiali/kiali-ui/pull/2056/commits/148ab1014d8acc9ce8f340023abb3f0bcb977732

so user can set 4XX or 4xx or [4|5]xx

the UI will replace it by 4\d\d or [4|5]\d\d

lucasponce commented:
@aljesusg I don't understand the last commit.

Please, take the time to review my comments, the idea is that user don't read by default the complex regex, but a friendly format to express 4xx and 5xx.

This has to be propagated in more places, operator included.

P.S.: Try also to rebase with master both backend + frontend as multicluster work has been merged

jshaughn reviewed (approved):
This seems to be working in the case I tried.  I did waste some time trying to figure out the latest syntax, so I look forward to some documentation to go with the feature.  a couple of very minor comments that you can fix, but approving.

lucasponce commented on models/health_config.go:
```diff
@@ -0,0 +1,26 @@
+	return []AnnotationKey{RateHealthAnnotation}
+}
+
+func GetHealthAnnotation(annotations map[string]string, filters []AnnotationKey) interface{} {
```
Perhaps "GetHealthAnnotation" should return a map[string]string instead of "interface{}".

Also, below there are multiple lines doing the conversion on the call:

```
GetHealthAnnotation(service.Annotations, GetHealthConfigAnnotation()).(map[string]string)
```

I guess if not found or if there is a type class it can return an error or a nil.

aljesusg replied:
I remember that i had problems in graph/telemetry/istio/appender/health_config.go when I return that to the metadata let me check

aljesusg replied:
mmm I did the commit with changes I am going to tests all again because I remember something about this

lucasponce commented on business/health.go:
```diff
@@ -188,6 +191,7 @@ func (in *HealthService) getNamespaceServiceHealth(namespace string, services []
 	// Prepare all data (note that it's important to provide data for all services, even those which may not have any health, for overview cards)
 	for _, service := range services {
 		h := models.EmptyServiceHealth()
+		h.Requests.HealthAnnotations = models.GetHealthAnnotation(service.Annotations, HealthAnnotation).(map[string]string)
```
I've commented in the "GetHealthAnnotation", it can return directly a "map[string]string" with nil or error if you want to check that the annotation is not well formed.

lucasponce commented on business/health.go:
```diff
@@ -311,12 +324,22 @@ func (in *HealthService) getAppRequestsHealth(namespace, app, rateInterval strin
+	if err != nil {
+		return rqHealth, err
+	}
+	if len(w.Pods) > 0 {
```
It seems now that it gets the annotation from the controller, instead of first Pod, it seems this check is not necessary.

lucasponce commented on business/services.go:
```diff
@@ -275,6 +276,26 @@ func (in *SvcService) GetService(namespace, service, interval string, queryTime
 	return &s, nil
 }
 
+func (in *SvcService) UpdateService(namespace, service string, interval string, queryTime time.Time, jsonPatch string) (*models.ServiceDetails, error) {
```
Note, in the current reduced scenario, Kiali UI is not updated the annotations.
Users typically can update directly the controller.

We can leave this method in the code as it's similar as the workload one, but edition of the annotations are out of the scope for the moment.

lucasponce commented on models/health_config.go:
```diff
@@ -0,0 +1,26 @@
+	return []AnnotationKey{RateHealthAnnotation}
+}
+
+func GetHealthAnnotation(annotations map[string]string, filters []AnnotationKey) interface{} {
```
Another comment, I looked close to this code.

I guess this signature:

```
func GetHealthAnnotation(annotations map[string]string, filters []AnnotationKey) interface{} {
```

Could be simplified as "filters" is always a constat in the current use case (if I looked the code correctly), so, I'd remove the "filters []AnnotationKey" and I'd use it internally of the method as a constant.

When this annotations feature grows, it can be refactored, but for now I guess it will help to simplify it.

lucasponce commented on graph/telemetry/istio/appender/health_config.go:
```diff
@@ -0,0 +1,59 @@
+		switch n.NodeType {
+		case graph.NodeTypeWorkload:
+			if workload, found := getWorkload(n.Workload, namespaceInfo); found {
+				n.Metadata[graph.HasHealthConfig] = models.GetHealthAnnotation(workload.HealthAnnotations, models.GetHealthConfigAnnotation())
```
Passing a "models.GetHealthConfigAnnotation()" to "models.GetHealthAnnotation" is weird, so I put a comment to use the filters as a constant in the package or method.

lucasponce reviewed (changes requested):
This time I've left some comments more related to code style and simplification.

No big blockers but some easy changes to address.

lucasponce reviewed (approved):
OK to consolidate.
As commented in the UI PR, no regressions, this feature is in active development and probably we will interate on it, but this is good.
